tests/test_send_recv_two_workers.py

```diff
@@ -0,0 +1,261 @@
+            assert obj == b"OK"
+            print("Shutting Down Server...")
+            ucp.stop_listener(li)
+            ucp.fin()
+
+        loop = asyncio.get_event_loop()
+        listener = ucp.start_listener(
+            write, listener_port=lisitener_port, is_coroutine=True
+        )
+        t = loop.create_task(listener.coroutine)
+        await t
+        print("Last Bit of Cleanup...")
+        ucp.fin()
+
+    loop = asyncio.get_event_loop()
+    loop.run_until_complete(f(port))
+
+
+def dataframe():
+    import cudf
+    import numpy as np
+
+    size = 2 ** 12
+    return cudf.DataFrame(
+        {"a": np.random.random(size), "b": np.random.random(size)},
+        index=np.random.randint(size, size=size),
+    )
+
+
+def column():
+    import cudf
+    return cudf.Series(np.arange(10000))._column
+
+
+def series():
+    import cudf
+    return cudf.Series(np.arange(10000))
+
+
+def empty_dataframe():
+    import cudf
+    return cudf.DataFrame({"a": [1.0], "b": [1.0]}).head(0)
+
+
+def cupy():
+    import cupy
+    return cupy.arange(10)
+
+
+def raise_error():
+    raise Exception
+
+
+@pytest.mark.parametrize(
+    "cuda_obj_generator", [dataframe, column, empty_dataframe, series, cupy, raise_error]
+)
+def test_send_recv_cudf(cuda_obj_generator):
+    base_env = {
+        "UCX_RNDV_SCHEME": "put_zcopy",
+        "UCX_MEMTYPE_CACHE": "n",
+        # "UCX_NET_DEVICES": "mlx5_0:1",
+        "UCX_TLS": "rc,cuda_copy,cuda_ipc",
+        "CUDA_VISIBLE_DEVICES": "0,1",
+    }
+    env1 = base_env.copy()
+    env2 = base_env.copy()
+    env2["CUDA_VISIBLE_DEVICES"] = "1,0"
+
+    port = 15338
+    # serialize function and send to the client and server
+    # server will use the return value of the contents,
+    # serialize the values, then send serialized values to client.
+    # client will compare return values the deserialize the
+    # data sent from the server
+
+    func = cloudpickle.dumps(cuda_obj_generator)
+
+    server_process = multiprocessing.Process(
+        name="server", target=server, args=[env1, port, func]
+    )
+    client_process = multiprocessing.Process(
+        name="client", target=client, args=[env2, port, func]
+    )
+
+    a = server_process.start()
+    b = client_process.start()
+
+    c = server_process.join()
+    d = client_process.join()
+
+    assert server_process.exitcode is 0
+    assert client_process.exitcode is 0
+
+
+def total_nvlink_transfer():
+    import pynvml
+
+    pynvml.nvmlInit()
+
+    try:
+        cuda_dev_id = int(os.environ["CUDA_VISIBLE_DEVICES"].split(",")[0])
+    except:
+        cuda_dev_id = 0
+    nlinks = pynvml.NVML_NVLINK_MAX_LINKS
+    # ngpus = pynvml.nvmlDeviceGetCount()
+    handle = pynvml.nvmlDeviceGetHandleByIndex(cuda_dev_id)
+    rx = 0
+    tx = 0
+    for i in range(nlinks):
+        transfer = pynvml.nvmlDeviceGetNvLinkUtilizationCounter(handle, i, 1)
+        rx += transfer["rx"]
+        tx += transfer["tx"]
+    return rx, tx
```

Comment:
We might want to do something like this as a pytest fixture to ensure that we're using nvlink

```python
@pytest.fixture
def uses_nvlink():
    # setup
    yield
    # test that transfer occurred
```

We would then include this within our tests above

```python
def test_foo(uses_nvlink):
```